src/math/exp_fcn_hifi4.c

```diff
@@ -0,0 +1,215 @@
+	ae_int32x2 in1_32 = AE_MOVINT32X2_FROMINT64(in_1);
+
+	ae_ep ep_lolo = AE_MOVEA(0);
+	ae_ep ep_hilo = AE_MOVEA(0);
+	ae_ep ep_HL_LH = AE_MOVEA(0);
+
+	producthihi = AE_MUL32_HH(in0_32, in1_32);
+	/* Input-32x32-bit(in1_32xin0_32)into 72-bit multiplication operations */
+	/* Output-lower 64 bits of the result are stored in producthilo */
+	/* Output-upper eight bits are stored in ep_hilo */
+	AE_MULZAAD32USEP_HL_LH(ep_hilo, producthilo, in1_32, in0_32);
+	productlolo = AE_MUL32U_LL(in0_32, in1_32);
+
+	product_hl_lh_h = AE_SRAI72(ep_hilo, producthilo, 32);
+	product_hl_lh_l = AE_SLAI64(producthilo, 32);
+
+	AE_ADD72(ep_lolo, productlolo, ep_HL_LH, product_hl_lh_l);
+
+	carry = AE_SRAI72(ep_lolo, productlolo, 32);
+
+	carry = AE_SRLI64(carry, 32);
+	producthi = AE_ADD64(producthihi, carry);
+
+	producthi = AE_ADD64(producthi, product_hl_lh_h);
+
+	*ptroutbitslo = productlolo;
+	*ptroutbitshi = producthi;
+}
+
+/*
+ * Arguments	: int64_t a
+ *		  int64_t b
+ * Return Type	: int64_t
+ */
+static int64_t mul_s64_sr_sat_near(int64_t a, int64_t b)
+{
+	ae_int64 result;
+	ae_int64 u64_chi;
+	ae_int64 u64_clo;
+	ae_int64 temp;
+
+	mul_s64(a, b, &u64_chi, &u64_clo);
+
+	ae_int64 roundup = AE_AND64(u64_clo, BIT_MASK_LOW_Q27P5);
+
+	roundup = AE_SRLI64(roundup, 27);
+	temp = AE_OR64(AE_SLAI64(u64_chi, 36), AE_SRLI64(u64_clo, 28));
+
+	return AE_ADD64(temp, roundup);
+}
+
+int64_t onebyfact_Q63[19] = {
+		4611686018427387904LL,
+		1537228672809129301LL,
+		384307168202282325LL,
+		76861433640456465LL,
+		12810238940076077LL,
+		1830034134296582LL,
+		228754266787072LL,
+		25417140754119LL,
+		2541714075411LL,
+		231064915946LL,
+		19255409662LL,
+		1481185358LL,
+		105798954LL,
+		7053264LL,
+		440829LL,
+		25931LL,
+		1441LL,
+		76LL,
+		4LL
+};
+
+/* f(x) = a^x, x is variable and a is base
+ *
+ * Arguments	: int32_t x(Q4.28)
+ * input range -5 to 5
+ *
+ * Return Type	: int32_t ts(Q9.23)
+ * output range 0.0067465305 to 148.4131488800
+ *+------------------+-----------------+--------+--------+
+ *| x		     | ts (returntype) |   x	|  ts	 |
+ *+----+-----+-------+----+----+-------+--------+--------+
+ *|WLen| FLen|Signbit|WLen|FLen|Signbit| Qformat| Qformat|
+ *+----+-----+-------+----+----+-------+--------+--------+
+ *| 32 | 28  |	1    | 32 | 23 |   0   | 4.28	| 9.23	 |
+ *+------------------+-----------------+--------+--------+
+ */
+int32_t sofm_hifi_exp_int32(int32_t x)
+{
+	ae_int64 outhi;
+	ae_int64 outlo;
+	ae_int64 qt;
+	ae_int64 onebyfact;
+	ae_int64 temp;
+
+	ae_int64 *ponebyfact_Q63 = (ae_int64 *)&onebyfact_Q63[0];
+	ae_int64 ts = TERMS_Q23P9;
+	ae_int64 mp = (x + LSHIFT_BITS) >> 14; /* x in Q50.14 */;
+	xtbool flag;
+
+	int64_t qt_temp;
+	int64_t b_n;
+
+	mul_s64(mp, BIT_MASK_Q62P2, &outhi, &outlo);
+	qt = AE_OR64(AE_SLAI64(outhi, 46), AE_SRLI64(outlo, 18));
+
+	temp = AE_SRAI64(AE_ADD64(qt, QUOTIENT_SCALE), 35);
+
+	ts = AE_ADD64(ts, temp);
+
+	mp = mul_s64_sr_sat_near(mp, (int64_t)x);
+
+	for (b_n = 0; b_n < 64;) {
+		ae_int64 sign;
```

Comment:
i think we are moving towards a c99 standard so this is no longer mandatory

Reply:
This is compatible with C89 because it's at the start of a `{` block.

Reminder: we're in 2023 :-)

C99 would be this:
```C
	for (b_n = 0; b_n < 64;) {

                 // some code
    
		const ae_int64 sign = ...

```